Raise bouncerscript level to a long-overdue production one.

MihaiTabara opened the pull request:
* 100% code-coverage
* docstrings
* fixes
* refactoring
* tasks, templates, examples, fixings

JohanLorenzo commented on bouncerscript/script.py:
```diff
@@ -34,11 +33,6 @@
             continue
 
         log.info("Adding {} ...".format(product_name))
-        # bug 1443104: temp hack that goes away when bug is solved
```
:+1:

JohanLorenzo commented on bouncerscript/script.py:
```diff
@@ -121,9 +110,7 @@ def main(name=None, config_path=None, close_loop=True):
-    if close_loop:
-        # Loop cannot be reopen once closed. Not closing it allows to run several tests on main()
-        loop.close()
+    loop.close()
```
Nit: you can remove `close_loop` from the function signature.

JohanLorenzo commented on bouncerscript/test/test_work_dir/task.json:
```diff
@@ -0,0 +1,35 @@
+    "payload": {
+        "locales": ["en-US", "ro"],
+        "submission_entries": {
+            "Firefox-99999": {
```
Nit: This might be good to have a second entry in the payload. This would ensure we do iterate on them

JohanLorenzo commented on bouncerscript/test/test_work_dir/task.json:
```diff
@@ -0,0 +1,35 @@
+{
```
Nit: It looks like this file is the same as `task_submission.json`. Should we keep both?

MihaiTabara replied:
https://github.com/mozilla-releng/scriptworker/blob/master/scriptworker/client.py#L42 expects `task.json` so unless we mock this function and overwrite the filename, the unit testing for `async_main` will barf. If we kill this, all the functions using any of the `submission_context` or `alias_context` will barf if we remote the others as they are loading the configurations.

I guess I can redo that part and make it nicer, but let's not block on that for now. I raise https://github.com/mozilla-releng/bouncerscript/issues/13 to track that separately.

JohanLorenzo replied:
Okay, I'll continue the conversation there :)

JohanLorenzo commented on bouncerscript/utils.py:
```diff
@@ -31,45 +32,43 @@ def load_json(path):
-            log.warning("Timed out accessing %s: %s" % (api_url, e))
-            raise
+
+    log.info("Submitting to %s" % api_url)
```
Nit: It seems the log line contains more information than this one. Should we get rid of this one? Should we make the other a `debug` one?

MihaiTabara replied:
Let's kill it, you're right, second one is much more verbose.

JohanLorenzo commented on bouncerscript/utils.py:
```diff
@@ -31,45 +32,43 @@ def load_json(path):
+                                                                       api_url,
+                                                                       kwargs))
+        async with session.request(method, api_url, auth=auth, **kwargs) as resp:
+            log.info("Server response")
```
Nit: This can be merged with `log.info(result)`. For instance: `log.info('Server response: {}'.format(result))

Nit 2: Should this be a `log.debug`?

MihaiTabara replied:
Used to be `log.info` in the old scripts too and with the recent issues we've had, I think it's safer to see more logs, hence keep `log.info()`. There aren't too many products anyway and logs are quite short anyway.

MihaiTabara replied:
But full agreed on merging the two, not sure how I missed that. Thanks!

JohanLorenzo commented on bouncerscript/utils.py:
```diff
@@ -31,45 +32,43 @@ def load_json(path):
+        raise
+    except aiohttp.ClientError as e:
+        log.warning("Cannot access %s" % api_url)
+        traceback.print_exc(file=sys.stdout)
```
Nit: What does this line do more than this one https://github.com/mozilla-releng/bouncerscript/pull/12/files#diff-835b68f16b034323ff73ec1722407044R110 ? Should we left the main function print the stack? Or is there some context I didn't get?

MihaiTabara replied:
You're absolutely right. I've killed in. To be honest, I bluntly copy-pasted some of the behaviors from the old script and forgot to revisit. Good catch.

JohanLorenzo commented on bouncerscript/utils.py:
```diff
@@ -31,45 +32,43 @@ def load_json(path):
+        log.warning("Timed out accessing %s: %s" % (api_url, e))
+        raise
+    except aiohttp.ClientError as e:
+        log.warning("Cannot access %s" % api_url)
```
Nit: `e` isn't logged out, but it is declared. Let's either remove `as e` or log it out.

JohanLorenzo commented on README.md:
```diff
@@ -1 +1,19 @@
+## deploy a new version
+
+1. Once your PR is reviewed and passes the tests, have one of the admins review & merge it
+2. Bump to new version in `version.json` and commit with a "%VERSION%" message
```
Nit: There is no `version.json`. `s/json/txt` ?

MihaiTabara replied:
Let's be consistent with `version.txt` then as I see most of *scripts are. I'll need to amend https://github.com/mozilla-releng/beetmoverscript/blob/master/version.json and https://github.com/mozilla-releng/balrogscript/blob/master/version.json at some point too.  Good catch again!

JohanLorenzo commented on README.md:
```diff
@@ -1 +1,19 @@
+7. Wait for that file to be synchronized across all puppet instances (emails arrive to confirm that)
+8. Tweak the `bouncerscript` version under [bouncerworker module](https://hg.mozilla.org/build/puppet/file/tip/modules/bouncer_scriptworker/manifests/init.pp#l29) to reflect the new value
+9. Push puppet bump to `default` branch, wait for tests to run and confirmation to arrive in `#releng`. Merge it to `production` after that.
+10. There is currently *a single instance of* bouncerworker. Ssh to it and run `puppet agent --test` as `root` to enforce the deployment of the newest catalog. Can also wait for the cron job to run puppet to deploy new changes every 30 mins or so.
```
Nit: `as root` can be replaced by `sudo puppet agent --test`. That removes the ambiguity of logging as root.

JohanLorenzo commented on README.md:
```diff
@@ -1 +1,19 @@
+5. `git push --tags`
+6. Create tarball with `python setup.py sdist` and scp that file under [puppet](http://releng-puppet2.srv.releng.scl3.mozilla.com/python/packages-3.5/)
+7. Wait for that file to be synchronized across all puppet instances (emails arrive to confirm that)
+8. Tweak the `bouncerscript` version under [bouncerworker module](https://hg.mozilla.org/build/puppet/file/tip/modules/bouncer_scriptworker/manifests/init.pp#l29) to reflect the new value
```
Nit: The line number may not be valid anymore of looking at the tip file. I'd recommend to `s/tip/some_known_rev`

MihaiTabara replied:
DXR permalink FTW https://dxr.mozilla.org/build-central/rev/a5b360575c6f6b67b1093b81145f4700b13bd9da/puppet/modules/bouncer_scriptworker/manifests/init.pp#29 👍

JohanLorenzo commented on README.md:
```diff
@@ -1 +1,19 @@
+[![Coverage Status](https://coveralls.io/repos/github/mozilla-releng/bouncerscript/badge.svg?branch=master)](https://coveralls.io/github/mozilla-releng/bouncerscript?branch=master)
+
+
+## deploy a new version
```
:+1: for this piece of documentation!

JohanLorenzo commented on bouncerscript/test/test_utils.py:
```diff
@@ -14,3 +36,198 @@ def test_load_json_from_file():
+    )
+
+    assert response == '{}'
+    context.session.close()
```
Question for my own curiosity: Would it be possible to let the fixture (https://github.com/mozilla-releng/bouncerscript/pull/12/files#diff-3d6daaf40d168d2ef7c9b7e2fa652ff7R48) close the context? I don't know if that's feasible

MihaiTabara replied:
I don't think you could since not all the `context` objects have a session defined.

JohanLorenzo commented on bouncerscript/utils.py:
```diff
@@ -31,45 +32,43 @@ def load_json(path):
+        raise
 
 
 async def product_exists(context, product_name):
```
Nit: We could make clearer that the function returns a boolean, by using an interrogative form. For instance: `does_product_exist(context, product_name)`

MihaiTabara commented:
Addressed all comments from @JohanLorenzo. Commented if I didn't. Merging this but still won't roll-out to production until we're fully testing this. Will follow-up with other PRs to:
* add pre-checks for the aliases we're updating bug 1445684
* factorize common stuff among *scripts bug 1445593

After enough testing, we'll roll-out to production the whatever version we'll have then.

JohanLorenzo commented:
\o/

MihaiTabara commented:
@JohanLorenzo Thanks for all the feedback by the way, much appreciated! \o/
